utils/symbol-rawelf.c

```diff
@@ -13,6 +13,86 @@
 #include "utils/utils.h"
 #include "utils/symbol-rawelf.h"
 
+/*
+ *  ELF File Header validation logic.
+ */
+int elf_validation(struct uftrace_elf_data *elf)
+{
+	Elf_Ehdr *ehdr;
+	int eclass, data, version;
+	unsigned long mapped_saddr, addr;
+
+	ehdr = &elf->ehdr;
+	elf->have_section = false;
+
+	// validate ELF Magic.
+	if (memcmp(ehdr, ELFMAG, SELFMAG)) {
+		pr_dbg3("ELF Signature not matched\n");
+		return -1;
+	}
+
+	// validate some field of elf header.
+	eclass = (int) ehdr->e_ident[EI_CLASS];
+	data = (int) ehdr->e_ident[EI_DATA];
+	version = (int) ehdr->e_ident[EI_VERSION];
+
+	if (!(eclass > ELFCLASSNONE && eclass < ELFCLASSNUM)) {
+		pr_dbg3("Invalid eclass : [%d]\n", eclass);
+		return -1;
+	}
+
+	if (!(data > ELFDATANONE && data < ELFDATANUM)) {
+		pr_dbg3("Invalid endian : [%d]\n", data);
+		return -1;
+	}
+
+	if (!(version > EV_NONE && version < EV_NUM)) {
+		pr_dbg3("Invalid ELF version : [%d]\n", version);
+		return -1;
+	}
+
+	if (ehdr->e_phnum == 0 || ehdr->e_phentsize == 0) {
+		pr_dbg3("Invalid Program header. Num:[%d] Size:[%d]\n",
+				ehdr->e_phnum, ehdr->e_phentsize);
+		return -1;
+	}
+
+	if (ehdr->e_shnum > 0 && ehdr->e_shentsize == 0) {
+		pr_dbg3("Section Header entry size cannot be 0.\n");
+		return -1;
+	}
+
+	// start address the ELF mapped.
+	mapped_saddr = (long)elf->file_map;
+	addr = mapped_saddr + ehdr->e_phoff +
+		ehdr->e_phnum * ehdr->e_phentsize;
+
+	if (addr > mapped_saddr + elf->file_size) {
+		pr_dbg3("Invalid Program Header address. "\
+				"mapped address : [%llx - %llx]" \
+				"but phdr address is : [%llx]\n",
+				mapped_saddr,
+				mapped_saddr + elf->file_size, addr);
```

Comment:
I don't think the address is meaningful.  Also `mapped_saddr` is not needed for the check - we can only check the offsets.